backend/lib/plugins/socket.js

```diff
@@ -0,0 +1,555 @@
+    }
+    socket.userId = userId;
+
+    this.io.emit("USER_STATUS_UPDATE", { id: socket.userId, status: "online" });
+    socket.on("disconnect", () => {
+      this.log.info(`[ws] socket disconnected [socketId: ${socket.id}]`);
+      for (let room in socket.rooms) {
+        socket.leave(room);
+      }
+      setTimeout(async () => {
+        // the user maybe is just changing/reloading the page, wait to make sure they completely disconnected
+        const stillOnline = await getSocketIdByUserId(this, socket.userId);
+        if (!stillOnline)
+          this.io.emit("USER_STATUS_UPDATE", {
+            id: socket.userId,
+            status: "offline",
+          });
+      }, 1000);
+    });
+    socket.join(userId); // to send events to all the user's browsers, if many are open.
+    res({ code: 200, data: userId });
+    this.log.info(
+      `[ws] socket identified [socketId: ${socket.id}] [userId: ${userId}]`,
+    );
+  });
+
+  socket.on("JOIN_ROOM", async (data, res) => {
+    userId = socket.userId;
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    var threadErr, thread;
+    if (data.threadId) {
+      // joining from inbox
+      [threadErr, thread] = await this.to(
+        Thread.findOne({
+          _id: data.threadId,
+          "participants.id": userId,
+        }),
+      );
+    } else if (data.receiverId) {
+      // joining from feed
+      [threadErr, thread] = await this.to(
+        Thread.findOne({
+          $and: [
+            { "participants.id": userId },
+            { "participants.id": data.receiverId },
+          ],
+        }),
+      );
+    } else {
+      // user called joinRoom({}) inside leaveAllRooms()
+      for (let room in socket.rooms) {
+        if (![socket.id, socket.userId].includes(room))
+          await socket.leave(room);
+      }
+      return res({ code: 401, message: "Unauthorized" });
+    }
+    if (threadErr) return res({ code: 500, message: "Internal server error" });
+
+    // first time joining, creating new thread.
+    if (!thread && data.receiverId) {
+      const [senderErr, sender] = await this.to(User.findById(userId));
+      const [receiverErr, receiver] = await this.to(
+        User.findById(data.receiverId),
+      );
+      if (senderErr || receiverErr)
+        return res({ code: 500, message: "Internal server error" });
+      if (!sender || !receiver)
+        return res({ code: 404, message: "User not found" });
+      let newThread = { participants: [] };
+      [sender, receiver].forEach((participant) => {
+        newThread.participants.push({
+          id: participant._id,
+          lastAccess: participant == sender ? new Date() : null,
+          name: participant.name,
+          newMessages: participant == receiver ? 1 : 0,
+          photo: participant.photo,
+          status: participant == sender ? "accepted" : "pending",
+          type: participant.type,
+        });
+      });
```

Comment:
Minor: could use map:
```suggestion
      const newThread = { 
        participants: [sender, receiver].map((participant) => ({
          id: participant._id,
          lastAccess: participant == sender ? new Date() : null,
          name: participant.name,
          newMessages: participant == receiver ? 1 : 0,
          photo: participant.photo,
          status: participant == sender ? "accepted" : "pending",
          type: participant.type,
        }) 
      };
```